src/urllib3/util/proxy.py

```diff
@@ -0,0 +1,79 @@
+import os
+
+from .ssl_ import (
+    resolve_cert_reqs,
+    resolve_ssl_version,
+    create_urllib3_context,
+)
+
+
+def connection_requires_http_tunnel(
+    proxy_url=None, proxy_config=None, destination_scheme=None
+):
+    """
+    Returns True if the connection requires an HTTP CONNECT through the proxy.
+
+    :param destination_url:
+        :URL URL of the destination.
+    :param proxy_url:
+        :URL URL of the proxy.
+    :param proxy_config:
+        :class:`PoolManager.ProxyConfig` proxy configuration
+    """
+    # If we're not using a proxy, no way to use a tunnel.
+    if proxy_url is None:
+        return False
+
+    # HTTP destinations never require tunneling, we always forward.
+    if destination_scheme == "http":
+        return False
+
+    # Support for forwarding with HTTPS proxies and HTTPS destinations.
+    if (
+        proxy_url.scheme == "https"
+        and proxy_config
+        and proxy_config.use_forwarding_for_https
+    ):
+        return False
+
+    # Otherwise always use a tunnel.
+    return True
+
+
+def generate_proxy_ssl_context(
+    ssl_version, cert_reqs, ca_certs=None, ca_cert_dir=None, ca_cert_data=None
+):
+    """
+    Generates a default proxy ssl context if one hasn't been provided by the
+    user.
+    """
+    ssl_context = create_urllib3_context(
+        ssl_version=resolve_ssl_version(ssl_version),
+        cert_reqs=resolve_cert_reqs(cert_reqs),
+    )
+    if (
+        not ca_certs
+        and not ca_cert_dir
+        and not ca_cert_data
+        and hasattr(ssl_context, "load_default_certs")
+    ):
+        ssl_context.load_default_certs()
+
+    proxy_cert, proxy_key, proxy_pass = client_certificate_and_key_from_env()
+
+    if proxy_cert:
+        ssl_context.load_cert_chain(proxy_cert, keyfile=proxy_key, password=proxy_pass)
+
+    return ssl_context
+
+
+def client_certificate_and_key_from_env():
+    """
+    Attempts to retrieve a client certificate and key from the environment
+    variables to use with the proxy.
+    """
+    proxy_cert = os.environ.get("PROXY_SSLCERT")
```

Comment:
So I'm kinda in -1 territory when it comes to environment variables. Ideally our users would make these sorts of things configurable with their own interfaces. We don't support setting connection TLS config w/ environment variables either.

Reply:
I agree with you in principle. Ideally all libraries should provide interfaces that allow clients to pass through a proxy ssl context and any necessary parameters to urllib3. I already have prepared upstream patches to do this for botocore and requests. I'll send those out once we release the changes in urllib3. 

That said though, there's a ton of libraries out there that would need clients to prepare PRs or changes to allow these parameters to go through. It seems easier to support those through environment variables for the time being.

Are you concerned about these being accidentally set? Could prefixing them by ```URLLIB3_*``` address that concern?

Reply:
Knowing when to use environment variables versus parameters in the program versus OS defaults is a big can of worms as we know from Requests, so I want to mark it out of scope for this PR and release. I can definitely see the value, I don't want to release it as a part of 1.26 though.

I appreciate that you have PRs for boto and requests though, that will be very useful! :)